src/components/MultiSelect/MultiSelect.tsx

```diff
@@ -1,138 +1,433 @@
+      const isIndeterminate = (
+        multiSelectId: string,
+        item: MultiSelectItem
+      ): boolean => {
+        let childIds: string[] = item.children.map((childItem) => childItem.id);
+        if (
+          selectedItems[multiSelectId] !== undefined &&
+          childIds.length > 0 &&
+          childIds.some(
+            (childItem) =>
+              selectedItems[multiSelectId].items.includes(childItem) &&
+              !item.children.find((c) => c.id === childItem)?.isDisabled
+          )
+        ) {
+          return !isAllChecked(multiSelectId, item);
+        }
+        return false;
+      };
+
+      const updateDisabledState = (
+        multiSelectId: string,
+        item: MultiSelectItem
+      ): boolean => {
+        // Update isDisabled for the parent item
+        item.isDisabled = isAllChecked(multiSelectId, item);
+
+        // Check if all child items are disabled, and update the parent accordingly
+        if (item.children.every((childItem) => childItem.isDisabled)) {
+          item.isDisabled = true;
+        } else {
+          item.isDisabled = false;
+        }
+        // Return the updated isDisabled state for the parent item
+        return item.isDisabled;
       };
 
-      if (type === "listbox") {
-        const listboxOnChange = onChange as ListboxOnChange;
+      const onChangeSearch = (event) => {
+        const value = event.target.value.trim().toLowerCase();
 
+        if (value === "") {
+          return listItemsCount === defaultItemsVisible
+            ? displayDefaultItems()
+            : setItemsList(items);
+        }
+
+        const filteredItems = items.filter((item) => {
+          if (item.children) {
+            if (item.name.toLowerCase().includes(value)) {
+              return true;
+            }
+
+            return item.children.some((childItem) =>
+              childItem.name.toLowerCase().includes(value)
+            );
+          }
+
+          return item.name.toLowerCase().includes(value);
+        });
+
+        setItemsList(filteredItems);
+      };
+
+      const clearSearchKeyword = () => {
+        if (listOverflow === "scroll") {
+          setItemsList(items);
+        } else {
+          listItemsCount === defaultItemsVisible
+            ? displayDefaultItems()
+            : setItemsList(items);
+        }
+      };
+
+      const displayDefaultItems = () => {
+        const list = [];
+        let count = 0;
+
+        for (let i = 0; i < items.length && count < defaultItemsVisible; i++) {
+          const currentItem = { ...items[i] };
+          count++;
+          if (currentItem.children) {
+            const showChilds = isAllChecked(id, currentItem);
+            const showIndeterminateChilds = isIndeterminate(id, currentItem);
+            count +=
+              !showChilds || !showIndeterminateChilds
+                ? currentItem.children?.length
+                : items[i].children.length;
```

Comment:
This is confusing. Both `currentItem` and `items[i]` are the same so regardless of what is happening, `count` will always get the length count which will be the same number.